### DISCUSSION TITLE
Extending scope of Carbon Aware SDK

### DISCUSSION THREAD

**Discussion by YaSuenag** [Green-Software-Foundation/carbon-aware-sdk#55] [General]
I'd like to propose extending the project scope of the SDK.

When I tried current version of SDK, I’m aware that the SDK is only focusing on gathering CO2 intensity (parameter `I` for SCI score) for each location, however I think it would not be enough to apply to production systems.

![sdk-components](https://user-images.githubusercontent.com/7421132/165662193-51e1b474-3c28-4764-8d22-300acbdd105f.png)
Fig. Carbon Aware SDK components from [project proposal](https://docs.google.com/document/d/1YG4HM3m3anLYbFZ-jl_h1LVxoV1yAcJkBGA40W8Rj-A) (GSF members only)

We should support recording SCI parameters and providing “current” CO2 intensity (`I`) for on-premise or cloud environment because we assume the SDK would be used to build pipeline, ML input data, and scheduler / router on Karbonates. This means the scope of Carbon Aware Core should include some components of “POSSIBLE GSF PROJECTS” (such as energy consumption, data providers, etc).

### Use case 1: Carbon-Aware scheduling

We want to deploy our apps on an environment with lower CO2 intensity (`I`) if we have multiple choice of location. Therefore, we need “current” CO2 intensity (`I`) values from reliable sources to determine which location we should deploy. This should include data from data center (on-premise facilities) since some data center has their own mechanisms to provide energy such as solar panel.  (this type is not supported , for example, by [electricityMap](https://app.electricitymap.org/map))

### Use case 2: Recording Energy Consumption Data

Eventually, we will use SCI score (`(E x I) + M) per R`) to evaluate whether our apps on data centers (including cloud environments) is carbon friendly. To achieve this, SCI requires the value of energy consumption, marginal carbon intensity, and embodied carbon emission. Though we cannot use them for the first deployment using SDK (as we never know E, R parameters before running application on Server), we should record data (`E`, `I`, `M`) in SDK to support future SCI calculations.

**Comment by YaSuenag** [comment]
I've sent PR for this discussion in #57 . I think this PR would cover this proposal.

If this proposal and PR is accepted, I want to add plugin implementation for on-premise - it supports measuring energy consumption.

**Comment by Sealjay** [comment]
This is really cool. I think the related projects we should discuss in the OSS WG, but the extension sounds like a great idea. FYI @dtoakley

**Reply by YaSuenag** [reply]
My colleague @yuu1127 explaind about this in last OSS WG (11 May, 2022), but we couldn't discuss for details at that time.

I think this proposal affects to entire of the SDK because it will change interfaces. So I want to discuss about this ASAP because other folks start to work the SDK (e.g. WattTime integration).

**Comment by Willmish** [comment]
I propose we reactivate this thread after v1.0 goes live, as this is an interesting proposition that needs some more thinking about from all of us.
